File: app/src/main/java/com/example/anxiao/mytestapplication/MainListAdapter.java
package com.example.anxiao.mytestapplication;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainListAdapter extends RecyclerView.Adapter<MainListAdapter.ListItemVH> {

    private LayoutInflater inflater;
    private List<MainListItemBean> mDates = new ArrayList<>();
    private MainListAction mAction;

    public MainListAdapter(Context context, MainListAction action) {
        this.inflater = LayoutInflater.from(context);
        this.mAction = action;
    }

    public void setDates(List<MainListItemBean> list) {
        if (list != null) {
            this.mDates = list;
            notifyDataSetChanged();
        }
    }

    @Override
    public ListItemVH onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ListItemVH(inflater.inflate(R.layout.main_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(ListItemVH holder, int position) {
        holder.itemDescribe.setText(String.valueOf(mDates.get(position).getDescribe()));
    }

    @Override
    public int getItemCount() {
        return mDates.size();
    }

    class ListItemVH extends RecyclerView.ViewHolder {
        @BindView(R.id.list_item_describe)
        TextView itemDescribe;

        public ListItemVH(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @OnClick(R.id.list_item_describe)
        void onClick() {
            mAction.itemTap(getAdapterPosition());
        }
    }

    public interface MainListAction {
        void itemTap(int index);
    }
}

File: app/src/main/java/com/example/testapp/MainListAdapter.kt
package com.example.testapp

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.example.testapp.mytestapplication.R

import java.util.ArrayList

class MainListAdapter(context: Context, private val mAction: MainListAction) : RecyclerView.Adapter<MainListAdapter.ListItemVH>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private var mDates = ArrayList<MainListItemBean>()

    fun setDates(list: ArrayList<MainListItemBean>) {
        this.mDates = list
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListItemVH {
        return ListItemVH(inflater.inflate(R.layout.main_list_item, parent, false))
    }

    override fun onBindViewHolder(holder: ListItemVH, position: Int) {

        holder.itemDescribe.text = mDates[position].describe
        holder.itemDescribe.setOnClickListener {
            mAction.itemTap(position)
        }

    }

    override fun getItemCount(): Int {
        return mDates.size
    }

    inner class ListItemVH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val itemDescribe = itemView.findViewById<TextView>(R.id.list_item_describe)
    }

    interface MainListAction {
        fun itemTap(index: Int)
    }
}
